widget/editor.go

```diff
@@ -223,8 +279,48 @@ func (e *Editor) processPointer(gtx layout.Context) {
 			if e.scroller.State() != gesture.StateFlinging {
 				e.caret.scroll = true
 			}
+
+			if evt.Modifiers == key.ModShift {
+				// If they clicked closer to the start, then change the start to
+				// where the caret used to be (effectively swapping start & end).
+				if abs(e.rr.caret-e.selStart.ofs) < abs(e.rr.caret-prevCaretPos.ofs) {
+					e.selStart = prevCaretPos
+				}
+			} else {
+				e.ClearSelection()
+			}
+			e.dragging = true
+
+			// Process a double-click.
+			if evt.NumClicks == 2 {
+				e.moveWord(-1, SelectionClear)
+				e.moveWord(1, SelectionExtend)
+				e.dragging = false
+			}
+		}
+	}
+
+	// e.dragger gets the same Release events as e.clicker, so if e.clicker
+	// processed a double-click event, we need to ignore the Release events
+	// here.
```

Comment:
This comment suggests a design problem in gestures, or even in pointer event resolution. Can you think of a way to change Dragger, or something else to support composing gestures better? I find the logic hard to follow.

Reply:
Yeah, the logic isn't the clearest.

I'm going to try something like

```go
for _, evt := range append(e.clicker.Events(gtx), e.dragger.Events(gtx.Metric, gtx, gesture.Both))
```

and see if that works and is clearer.

Reply:
Well, something similar works:

```go
	for _, evt := range e.clickDragEvents(gtx) {
		switch evt := evt.(type) {
		case gesture.ClickEvent:
                // ...
		case pointer.Event:
                // ...
		}
	}
       // ...

func (e *Editor) clickDragEvents(gtx layout.Context) []event.Event {
	var combinedEvents []event.Event
	for _, evt := range e.clicker.Events(gtx) {
		combinedEvents = append(combinedEvents, evt)
	}
	for _, evt := range e.dragger.Events(gtx.Metric, gtx, gesture.Both) {
		combinedEvents = append(combinedEvents, evt)
	}
	return combinedEvents
}